text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+The `configureDependencies` hook is the _only_ way to disable child packages. The package hooks are implemented as a class with no base class. There is no `super` to manipulate to interfere with your children’s hooks.
+
+### Build Hook: skipBabel
+
+```ts
+skipBabel(): { package: string, semverRange?: string }[]
+```
+
+By default, all imported dependencies (and their recursive imported dependencies) go through the app's babel config. This ensures browser compatibility safety. However, we provide `skipBabel` as an opt-out to work around transpilation problems in cases where the developer has verified that transpilation of a given package isn't needed.
+
+`skipBabel` returns a list of package names and optionally semver ranges. If no range is included, it defaults to `*`. This is a place where you're allowed to mentioned packages that are _not_ in your **allowed dependencies** because it may be necessary to talk about deeper dependencies within them. The `skipBabel` settings for all active addons are combined and if any addon skips babel for a given package & version, that causes the package to not be transpiled.
+
+The semver range is useful to disambiguate if there are multiple versions of the same package involved in the app, and in cases where a developer has manually verified that transpilation isn't needed, it's good practice to use the semver range so that `skipBabel` doesn't accidentally apply to a future version of the package that may indeed need transpilation.
+
+## What about Test Support?
+
+v1 packages can provide `treeForTestSupport`, `treeForAddonTestSupport`, and `app.import` with `type="test"`. All of these features are dropped.
+
+To provide test-support code, make a separate module within your package and tell people to `import` it from their tests. As long as it is only imported from tests, it will not be present in non-test bundles. (Things get simpler when you respect the module dependency graph.)
+
+## Macro System
+
+v1 packages can run arbitrary Node code that completely alters their runtime code. This makes them impossible to analyze. v2 packages are not allowed to do this. There are no "treeFor\*" hooks. Instead, they can influence their runtime code only through the macro system.
+
+It helps to think about the macro system as an extension to Javascript itself that we allow in v2 packages. Because the macros are allowed to appear in any published V2 package, and because the macros are _not_ a dependency that each package can control (you don't get to bring your own separate macro system version with you), it's important that we design a small core that we can support for the long-term. We probably can't make breaking changes to the macro system, we can only introduce new macros.
```

Comment:
This was maybe more paranoid than necessary, just because a lot of the existing build tooling out there is bad at per-package configuration. For example, telling babel (a *single* babel, which is important!) to run different plugins for different packages requires a lot of bring-your-own infrastructure to make that work, where you end up doing relatively expensive searches to reverse engineer which package owns which file.

That said, it turns out we already needed that infrastructure and we do have it. So we could choose to use the `ember.version` key that we discussed elsewhere to influence the meaning of the macros if we really need to in the future.

I think it's fine to keep the RFC in this more conservative position, and it doesn't foreclose the ability to do those things if we need them in the future.

Reply:
I remembered the biggest reason why you don't get to bring your own version of the macros: because their implementation is tied to a specific version of babel (or ember-cli-htmlbars). That has caused big pain in the ecosystem in the past.

For example, babel 8 is coming. As long as we port embroider/macros to babel 8, an app using all v2 addons will be free to upgrade to babel 8 unilaterally, with no changes in its addons. Whereas if we allow addons to bring their own macro implementations, you can't upgrade until the lowest common denominator addon has done the work to upgrade their potentially-idiosyncratic macro implementation.

We could lift this restriction in the future by inventing a more declarative API for defining macros.